lib/matplotlib/colorbar.py

```diff
@@ -510,166 +490,276 @@ def __init__(self, ax, cmap=None,
         self.update_ticks()
+
         if self.filled:
-            self._add_solids(X, Y, self._values[:, np.newaxis])
+            ind = np.arange(len(self._values))
+            if self._extend_lower():
+                ind = ind[1:]
+            if self._extend_upper():
+                ind = ind[:-1]
+            self._add_solids(X, Y, self._values[ind, np.newaxis])
 
-    def _config_axis(self):
-        """Set up long and short axis."""
-        ax = self.ax
+    def _add_solids(self, X, Y, C):
+        """Draw the colors; optionally add separators."""
+        # Cleanup previously set artists.
+        if self.solids is not None:
+            self.solids.remove()
+        for solid in self.solids_patches:
+            solid.remove()
+        # Add new artist(s), based on mappable type.  Use individual patches if
+        # hatching is needed, pcolormesh otherwise.
+        mappable = getattr(self, 'mappable', None)
+        if (isinstance(mappable, contour.ContourSet)
+                and any(hatch is not None for hatch in mappable.hatches)):
+            self._add_solids_patches(X, Y, C, mappable)
+        else:
+            self.solids = self.ax.pcolormesh(
+                X, Y, C, cmap=self.cmap, norm=self.norm, alpha=self.alpha,
+                edgecolors='none', shading='flat')
+            if not self.drawedges:
+                if len(self._y) >= self.n_rasterize:
+                    self.solids.set_rasterized(True)
+        self.dividers.set_segments(
+            np.dstack([X, Y])[1:-1] if self.drawedges else [])
+
+    def _add_solids_patches(self, X, Y, C, mappable):
+        hatches = mappable.hatches * len(C)  # Have enough hatches.
+        patches = []
+        for i in range(len(X) - 1):
+            xy = np.array([[X[i, 0], Y[i, 0]],
+                           [X[i, 1], Y[i, 0]],
+                           [X[i + 1, 1], Y[i + 1, 0]],
+                           [X[i + 1, 0], Y[i + 1, 1]]])
+            patch = mpatches.PathPatch(mpath.Path(xy),
+                                       facecolor=self.cmap(self.norm(C[i][0])),
+                                       hatch=hatches[i], linewidth=0,
+                                       antialiased=False, alpha=self.alpha)
+            self.ax.add_patch(patch)
+            patches.append(patch)
+        self.solids_patches = patches
+
+    def _do_extends(self, extendlen):
+        """
+        Make adjustments of the inner axes for the extend triangles (or
+        rectanges) and add them as patches.
+        """
+        # extend lengths are fraction of the *inner* part of colorbar,
+        # not the total colorbar:
+        elower = extendlen[0] if self._extend_lower() else 0
+        eupper = extendlen[1] if self._extend_upper() else 0
+        tot = eupper + elower + 1
+        elower = elower / tot
+        eupper = eupper / tot
+        width = 1 / tot
+
+        bounds = np.array([0.0, elower, 1.0, width])
```

Comment:
Readability suggestions:

- Use "length" instead of "width".
- Don't make and use the "bounds" array until you absolutely need to; things like "top" and 
"xyout" will be more understandable in terms of "elower" and "length" than indexed entries in "bounds".